instrumentation/vertx/vertx-sql-client-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vertx/v4_0/sql/QueryResultBuilderInstrumentation.java

```diff
@@ -0,0 +1,98 @@
+import net.bytebuddy.matcher.ElementMatcher;
+
+public class QueryResultBuilderInstrumentation implements TypeInstrumentation {
+
+  @Override
+  public ElementMatcher<TypeDescription> typeMatcher() {
+    return named("io.vertx.sqlclient.impl.QueryResultBuilder");
+  }
+
+  @Override
+  public void transform(TypeTransformer transformer) {
+    transformer.applyAdviceToMethod(
+        named("tryComplete"),
+        QueryResultBuilderInstrumentation.class.getName() + "$CompleteAdvice");
+    transformer.applyAdviceToMethod(
+        named("tryFail").and(takesArguments(Throwable.class)),
+        QueryResultBuilderInstrumentation.class.getName() + "$FailAdvice");
+  }
+
+  @SuppressWarnings("unused")
+  public static class CompleteAdvice {
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static Scope onEnter(@Advice.FieldValue("context") Context vertxContext) {
+      ContextInternal contextInternal = (ContextInternal) vertxContext;
+      VertxSqlClientRequest otelRequest =
+          (VertxSqlClientRequest) contextInternal.localContextData().get(OTEL_REQUEST_KEY);
+      io.opentelemetry.context.Context otelContext =
+          (io.opentelemetry.context.Context)
+              contextInternal.localContextData().get(OTEL_CONTEXT_KEY);
+      io.opentelemetry.context.Context otelParentContext =
+          (io.opentelemetry.context.Context)
+              contextInternal.localContextData().get(OTEL_PARENT_CONTEXT_KEY);
+      if (otelRequest == null || otelContext == null || otelParentContext == null) {
+        return null;
+      }
+      instrumenter().end(otelContext, otelRequest, null, null);
+      return otelParentContext.makeCurrent();
+    }
+
+    @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
+    public static void onExit(@Advice.Enter Scope scope) {
+      if (scope != null) {
+        scope.close();
+      }
+    }
+  }
+
+  @SuppressWarnings("unused")
+  public static class FailAdvice {
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static Scope onEnter(
+        @Advice.Argument(0) Throwable throwable,
+        @Advice.FieldValue("context") Context vertxContext) {
+      ContextInternal contextInternal = (ContextInternal) vertxContext;
+      VertxSqlClientRequest otelRequest =
+          (VertxSqlClientRequest) contextInternal.localContextData().get(OTEL_REQUEST_KEY);
+      io.opentelemetry.context.Context otelContext =
+          (io.opentelemetry.context.Context)
+              contextInternal.localContextData().get(OTEL_CONTEXT_KEY);
+      io.opentelemetry.context.Context otelParentContext =
+          (io.opentelemetry.context.Context)
+              contextInternal.localContextData().get(OTEL_PARENT_CONTEXT_KEY);
+      if (otelRequest == null || otelContext == null || otelParentContext == null) {
+        return null;
+      }
+      instrumenter().end(otelContext, otelRequest, null, throwable);
+      return otelParentContext.makeCurrent();
```

Comment:
With the exception of the `throwable`, this is duplicated from above. How about a static method that takes the `Context` and a nullable `Throwable` and does this work for both places?